File: CS/E1887/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Design;
using DevExpress.Data;
using DevExpress.XtraTreeList.Columns;

namespace E1887 {
    public partial class Form1 : XtraForm {
        public Form1() {
            InitializeComponent();
        }

        private void OnShowExpressionEditorClick(object sender, EventArgs e) {
            using (ExpressionEditorForm editor = 
                new ConditionExpressionEditorForm(new TreeListDataColumnInfo(treeList1.Columns), null)) {
                if (editor.ShowDialog() == DialogResult.OK) {
                    using (InputForm input = new InputForm()) {
                        input.ShowDialog();
                        xpView1.AddProperty(input.ColumnName, editor.Expression);
                        treeList1.PopulateColumns();
                    }
                }
            }
        }

        private void OnFormLoad(object sender, EventArgs e) {
            treeList1.DataSource = xpView1;
        }
    }

    public class TreeListDataColumnInfo :IDataColumnInfo {
        private TreeListColumnCollection columns;
        private TreeListColumn column;

        public TreeListDataColumnInfo(TreeListColumnCollection columns) {
            this.columns = columns;
        }

        private TreeListDataColumnInfo(TreeListColumn column) {
            this.column = column;
        }

        #region IDataColumnInfo Members

        public string Caption {
            get {
                if (column == null) return string.Empty;
                return string.IsNullOrEmpty(column.Caption) ? column.FieldName : column.Caption;
            }
        }

        public List<IDataColumnInfo> Columns { get { return GetColumns(); } }

        private List<IDataColumnInfo> GetColumns() {
            if (column != null) return null;
            List<IDataColumnInfo> result = new List<IDataColumnInfo>();
            foreach (TreeListColumn col in columns)
                result.Add(new TreeListDataColumnInfo(col));
            return result;
        }

        public DataControllerBase Controller { get { return null; } }

        public string FieldName { get { return column == null ? string.Empty : column.FieldName; } }

        public Type FieldType { get { return column == null ? null : column.ColumnType; } }

        public string Name { get { return Caption; } }

        public string UnboundExpression { get { return FieldName; } }

        #endregion
    }
}

File: VB/E1887/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Design
Imports DevExpress.Data
Imports DevExpress.XtraTreeList.Columns

Namespace E1887

    Public Partial Class Form1
        Inherits XtraForm

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub OnShowExpressionEditorClick(ByVal sender As Object, ByVal e As EventArgs)
            Using editor As ExpressionEditorForm = New ConditionExpressionEditorForm(New TreeListDataColumnInfo(treeList1.Columns), Nothing)
                If editor.ShowDialog() = DialogResult.OK Then
                    Using input As InputForm = New InputForm()
                        input.ShowDialog()
                        xpView1.AddProperty(input.ColumnName, editor.Expression)
                        treeList1.PopulateColumns()
                    End Using
                End If
            End Using
        End Sub

        Private Sub OnFormLoad(ByVal sender As Object, ByVal e As EventArgs)
            treeList1.DataSource = xpView1
        End Sub
    End Class

    Public Class TreeListDataColumnInfo
        Implements IDataColumnInfo

        Private columnsField As TreeListColumnCollection

        Private column As TreeListColumn

        Public Sub New(ByVal columns As TreeListColumnCollection)
            columnsField = columns
        End Sub

        Private Sub New(ByVal column As TreeListColumn)
            Me.column = column
        End Sub

#Region "IDataColumnInfo Members"
        Public ReadOnly Property Caption As String Implements IDataColumnInfo.Caption
            Get
                If column Is Nothing Then Return String.Empty
                Return If(String.IsNullOrEmpty(column.Caption), column.FieldName, column.Caption)
            End Get
        End Property

        Public ReadOnly Property Columns As List(Of IDataColumnInfo) Implements IDataColumnInfo.Columns
            Get
                Return GetColumns()
            End Get
        End Property

        Private Function GetColumns() As List(Of IDataColumnInfo)
            If column IsNot Nothing Then Return Nothing
            Dim result As List(Of IDataColumnInfo) = New List(Of IDataColumnInfo)()
            For Each col As TreeListColumn In columnsField
                result.Add(New TreeListDataColumnInfo(col))
            Next

            Return result
        End Function

        Public ReadOnly Property Controller As DataControllerBase Implements IDataColumnInfo.Controller
            Get
                Return Nothing
            End Get
        End Property

        Public ReadOnly Property FieldName As String Implements IDataColumnInfo.FieldName
            Get
                Return If(column Is Nothing, String.Empty, column.FieldName)
            End Get
        End Property

        Public ReadOnly Property FieldType As Type Implements IDataColumnInfo.FieldType
            Get
                Return If(column Is Nothing, Nothing, column.ColumnType)
            End Get
        End Property

        Public ReadOnly Property Name As String Implements IDataColumnInfo.Name
            Get
                Return Caption
            End Get
        End Property

        Public ReadOnly Property UnboundExpression As String Implements IDataColumnInfo.UnboundExpression
            Get
                Return FieldName
            End Get
        End Property
#End Region
    End Class
End Namespace
